packages/rps/src/redux/game-v2/reducer.ts

```diff
@@ -0,0 +1,168 @@
+      return handleResign(state, action);
+    default:
+      return unreachable(action, state);
+  }
+};
+
+export const gameReducer = combineReducers({
+  localState: localReducer,
+  channelState: channelReducer,
+});
+
+const handleJoinOpenGame = (state: LocalState, action: JoinOpenGame): LocalState => {
+  if (state.type === 'Empty') {
+    return state;
+  }
+
+  const {opponentName, opponentAddress, roundBuyIn} = action;
+  const {name, address} = state;
+
+  return gameChosen({name, address, opponentName, roundBuyIn}, opponentAddress);
+};
+
+const handleGameJoined = (state: LocalState, action: GameJoined): LocalState => {
+  if (state.type !== 'WaitingRoom') {
+    return state;
+  }
+
+  const {opponentName, opponentAddress} = action;
+  const {name, address, roundBuyIn} = state;
+
+  return opponentJoined({name, address, opponentName, roundBuyIn, opponentAddress});
+};
+
+const handleCreateGame = (state: LocalState, action: CreateGame): LocalState => {
+  if (state.type === 'Empty') {
+    return state;
+  }
+
+  const {roundBuyIn} = action;
+  const {name, address} = state;
+
+  return waitingRoom({name, address, roundBuyIn});
+};
+
+const handleChooseWeapon = (state: LocalState, action: ChooseWeapon): LocalState => {
+  if (state.type !== 'ChooseWeapon') {
+    return state;
+  }
+
+  const {weapon} = action;
+  return weaponChosen(state, weapon);
+};
+
+const handleChooseSalt = (state: LocalState, action: ChooseSalt): LocalState => {
+  if (state.type !== 'WeaponChosen' || state.player !== 'A') {
+    return state;
+  }
+  // not sure why typsecript can't see that player === 'A' here ...
+  const oldLocalState = state as (WeaponChosen & {player: 'A'});
+
+  const {salt} = action;
+  return weaponAndSaltChosen(oldLocalState, salt);
+};
+
+const handleResultArrived = (state: LocalState, action: ResultArrived): LocalState => {
+  if (state.type !== 'WeaponChosen' && state.type !== 'WeaponAndSaltChosen') {
+    return state;
+  }
+  const {theirWeapon, result, fundingSituation} = action;
+  switch (fundingSituation) {
+    case 'Ok':
+      return resultPlayAgain(state, theirWeapon, result);
+    case 'MyFundsTooLow':
+      return shuttingDown(state, 'InsufficientFundsYou');
+    case 'OpponentsFundsTooLow':
+      return shuttingDown(state, 'InsufficientFundsOpponent');
+    default:
+      return state;
+  }
+};
+
+const handlePlayAgain = (state: LocalState, action: PlayAgain): LocalState => {
+  if (state.type !== 'ResultPlayAgain') {
+    return state;
+  }
+  return waitForRestart(state);
+};
+
+const handleRestart = (state: LocalState, action: Restart): LocalState => {
+  if (state.type !== 'WaitForRestart') {
+    return state;
+  }
+
+  return chooseWeapon(state);
+};
+
+const handleResign = (state: LocalState, action: Resign): LocalState => {
+  if (state.type === 'Empty' || state.type === 'Lobby' || state.type === 'WaitingRoom') {
+    return state;
+  }
+
+  return shuttingDown(state, 'YouResigned');
+};
```

Comment:
Just wanted to note that previous reducer was 883 lines. Your new reducer and saga combined are 168+191=359 lines. A rough metric, and the tests are incomplete: but seems encouraging that this is a cleaner implementation.